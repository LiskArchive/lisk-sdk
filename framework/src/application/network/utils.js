/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const net = require('net');
const dns = require('dns');

const lookupPromise = (hostname, options) =>
	new Promise((resolve, reject) => {
		dns.lookup(hostname, options, (err, address) => {
			if (err) {
				return reject(err);
			}

			return resolve(address);
		});
	});

const lookupPeersIPs = async (peersList, enabled) => {
	// If peers layer is not enabled there is no need to create the peer's list
	if (!enabled) {
		return [];
	}

	// In case domain names are used, resolve those to IP addresses.
	return Promise.all(
		peersList.map(async peer => {
			if (net.isIPv4(peer.ip)) {
				return peer;
			}

			try {
				const address = await lookupPromise(peer.ip, { family: 4 });
				return {
					...peer,
					ip: address,
				};
			} catch (err) {
				console.error(
					`Failed to resolve peer domain name ${peer.ip} to an IP address`,
				);
				return peer;
			}
		}),
	);
};

module.exports = {
	lookupPeersIPs,
};
