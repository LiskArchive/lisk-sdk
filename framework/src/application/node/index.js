/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { config: DefaultConfig } = require('./defaults');
const Chain = require('./chain');
const { migrations } = require('./components');
const BaseModule = require('../../modules/base_module');

/* eslint-disable class-methods-use-this */

class ChainModule extends BaseModule {
	constructor(options) {
		super(options);

		this.chain = null;
	}

	static get alias() {
		return 'chain';
	}

	static get info() {
		return {
			author: 'LiskHQ',
			version: '0.1.0',
			name: 'lisk-framework-chain',
		};
	}

	static get migrations() {
		return migrations;
	}

	static get defaults() {
		return DefaultConfig;
	}

	get events() {
		return [
			'bootstrap',
			'blocks:change',
			'transactions:confirmed:change',
			'signature:change',
			'transactions:change',
			'rounds:change',
			'multisignatures:signature:change',
			'multisignatures:change',
			'delegates:fork',
			'loader:sync',
			'dapps:change',
			'rebuild',
			'processor:sync',
			'processor:deleteBlock',
			'processor:broadcast',
			'processor:newBlock',
		];
	}

	get actions() {
		return {
			calculateSupply: {
				handler: action => this.chain.actions.calculateSupply(action),
			},
			calculateMilestone: {
				handler: action => this.chain.actions.calculateMilestone(action),
			},
			calculateReward: {
				handler: action => this.chain.actions.calculateReward(action),
			},
			getForgerPublicKeysForRound: {
				handler: async action =>
					this.chain.actions.getForgerPublicKeysForRound(action),
			},
			updateForgingStatus: {
				handler: async action => this.chain.actions.updateForgingStatus(action),
			},
			postSignature: {
				handler: async action => this.chain.actions.postSignature(action),
			},
			getForgingStatusForAllDelegates: {
				handler: async () =>
					this.chain.actions.getForgingStatusForAllDelegates(),
			},
			getTransactionsFromPool: {
				handler: async action =>
					this.chain.actions.getTransactionsFromPool(action),
			},
			getTransactions: {
				handler: async action => this.chain.actions.getTransactions(action),
				isPublic: true,
			},
			getSignatures: {
				handler: async () => this.chain.actions.getSignatures(),
				isPublic: true,
			},
			postTransaction: {
				handler: async action => this.chain.actions.postTransaction(action),
			},
			getSlotNumber: {
				handler: async action => this.chain.actions.getSlotNumber(action),
			},
			calcSlotRound: {
				handler: async action => this.chain.actions.calcSlotRound(action),
			},
			getNodeStatus: {
				handler: async () => this.chain.actions.getNodeStatus(),
			},
			getLastBlock: {
				handler: async () => this.chain.actions.getLastBlock(),
				isPublic: true,
			},
			getBlocksFromId: {
				handler: async action => this.chain.actions.getBlocksFromId(action),
				isPublic: true,
			},
			getHighestCommonBlock: {
				handler: async action =>
					this.chain.actions.getHighestCommonBlock(action),
				isPublic: true,
			},
		};
	}

	async load(channel) {
		this.chain = new Chain(channel, this.options);
		await this.chain.bootstrap();
		channel.publish('chain:bootstrap');
	}

	async unload() {
		return this.chain.cleanup();
	}
}

module.exports = ChainModule;
