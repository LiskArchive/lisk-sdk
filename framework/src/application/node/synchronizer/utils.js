/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { maxBy } = require('lodash');
const { ForkStatus } = require('@liskhq/lisk-bft');

const restoreBlocks = async (blocksModule, processorModule, tx = null) => {
	const tempBlocks = await blocksModule.dataAccess.getTempBlocks(
		{},
		{ sort: 'height:asc', limit: null },
		tx,
	);

	if (!tempBlocks || tempBlocks.length === 0) {
		return false;
	}

	for (const tempBlockEntry of tempBlocks) {
		const tempBlockInstance = await processorModule.deserialize(
			tempBlockEntry.fullBlock,
		);
		await processorModule.processValidated(tempBlockInstance, {
			removeFromTempTable: true,
		});
	}

	return true;
};

const clearBlocksTempTable = storageModule =>
	storageModule.entities.TempBlock.truncate();

const deleteBlocksAfterHeight = async (
	processorModule,
	blocksModule,
	logger,
	desiredHeight,
	backup = false,
) => {
	let { height: currentHeight } = blocksModule.lastBlock;
	logger.debug(
		{ desiredHeight, lastBlockHeight: currentHeight },
		'Deleting blocks after height',
	);
	while (desiredHeight < currentHeight) {
		logger.trace(
			{
				height: blocksModule.lastBlock.height,
				blockId: blocksModule.lastBlock.id,
			},
			'Deleting block and backing it up to temporary table',
		);
		const lastBlock = await processorModule.deleteLastBlock({
			saveTempBlock: backup,
		});
		currentHeight = lastBlock.height;
	}
};

/**
 * Allows to restore blocks if there are blocks left upon startup in temp_blocks table (e.g. node crashed)
 * Depends upon fork choice rule if blocks will be applied
 *
 * 1. Gets all blocks from temp_blocks table
 * 2. Sort blocks according to height as we want to lowest height (the next block to be applied)
 * 3. Uses next block with fork choice rule - if fork status indicates we should switch to different chain
 * we continue applying blocks using the `restoreBlocks` function.
 * Otherwise we truncate the temp_blocks table.
 */
const restoreBlocksUponStartup = async (
	logger,
	blocksModule,
	processorModule,
	storageModule,
) => {
	// Get all blocks and find lowest height (next one to be applied)
	const tempBlocks = await storageModule.entities.TempBlock.get(
		{},
		{
			sort: 'height:asc',
			limit: null,
		},
	);
	const blockLowestHeight = tempBlocks[0];
	const blockHighestHeight = tempBlocks[tempBlocks.length - 1];

	const nextTempBlock = await processorModule.deserialize(
		blockHighestHeight.fullBlock,
	);
	const forkStatus = await processorModule.forkStatus(nextTempBlock);
	const blockHasPriority =
		forkStatus === ForkStatus.DIFFERENT_CHAIN ||
		forkStatus === ForkStatus.VALID_BLOCK;

	// Block in the temp table has preference over current tip of the chain
	if (blockHasPriority) {
		logger.info('Restoring blocks from temporary table');
		await deleteBlocksAfterHeight(
			processorModule,
			blocksModule,
			logger,
			blockLowestHeight.height - 1,
			false,
		);
		// In case fork status is DIFFERENT_CHAIN - try to apply blocks from temp_blocks table
		await restoreBlocks(blocksModule, processorModule);
		logger.info('Chain successfully restored');
	} else {
		// Not different chain - Delete remaining blocks from temp_blocks table
		await clearBlocksTempTable(storageModule);
	}
};

const computeBlockHeightsList = (
	finalizedHeight,
	activeDelegates,
	listSizeLimit,
	currentRound,
) => {
	const startingHeight = Math.max(1, (currentRound - 1) * activeDelegates);
	const heightList = new Array(listSizeLimit)
		.fill(0)
		.map((_, i) => startingHeight - i * activeDelegates)
		.filter(height => height > 0);
	const heightListAfterFinalized = heightList.filter(
		height => height > finalizedHeight,
	);
	return heightList.length !== heightListAfterFinalized.length
		? [...heightListAfterFinalized, finalizedHeight]
		: heightListAfterFinalized;
};

// eslint-disable-next-line class-methods-use-this
const computeLargestSubsetMaxBy = (arrayOfObjects, propertySelectorFunc) => {
	const maximumBy = maxBy(arrayOfObjects, propertySelectorFunc);
	const absoluteMax = propertySelectorFunc(maximumBy);
	const largestSubset = [];
	// eslint-disable-next-line no-restricted-syntax
	for (const item of arrayOfObjects) {
		if (propertySelectorFunc(item) === absoluteMax) {
			largestSubset.push(item);
		}
	}
	return largestSubset;
};

module.exports = {
	computeBlockHeightsList,
	computeLargestSubsetMaxBy,
	deleteBlocksAfterHeight,
	restoreBlocksUponStartup,
	restoreBlocks,
	clearBlocksTempTable,
};
