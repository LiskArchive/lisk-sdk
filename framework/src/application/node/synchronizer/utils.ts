/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { ForkStatus, BFT } from '@liskhq/lisk-bft';
import { Chain } from '@liskhq/lisk-chain';
import { Processor } from '../processor';
import { Logger } from '../../logger';

export const restoreBlocks = async (
	chainModule: Chain,
	processorModule: Processor,
): Promise<boolean> => {
	const tempBlocks = await chainModule.dataAccess.getTempBlocks();

	if (tempBlocks.length === 0) {
		return false;
	}

	for (const tempBlock of tempBlocks.reverse()) {
		await processorModule.processValidated(tempBlock, {
			removeFromTempTable: true,
		});
	}

	return true;
};

export const clearBlocksTempTable = async (chainModule: Chain): Promise<void> =>
	chainModule.dataAccess.clearTempBlocks();

export const deleteBlocksAfterHeight = async (
	processorModule: Processor,
	chainModule: Chain,
	logger: Logger,
	desiredHeight: number,
	backup = false,
): Promise<void> => {
	let { height: currentHeight } = chainModule.lastBlock.header;
	logger.debug({ desiredHeight, lastBlockHeight: currentHeight }, 'Deleting blocks after height');
	while (desiredHeight < currentHeight) {
		logger.trace(
			{
				height: chainModule.lastBlock.header.height,
				blockId: chainModule.lastBlock.header.id,
			},
			'Deleting block and backing it up to temporary table',
		);
		await processorModule.deleteLastBlock({
			saveTempBlock: backup,
		});
		currentHeight = chainModule.lastBlock.header.height;
	}
};

/**
 * Allows to restore blocks if there are blocks left upon startup in temp_blocks table (e.g. node crashed)
 * Depends upon fork choice rule if blocks will be applied
 *
 * 1. Gets all blocks from temp_blocks table
 * 2. Sort blocks according to height as we want to lowest height (the next block to be applied)
 * 3. Uses next block with fork choice rule - if fork status indicates we should switch to different chain
 * we continue applying blocks using the `restoreBlocks` function.
 * Otherwise we truncate the temp_blocks table.
 */
export const restoreBlocksUponStartup = async (
	logger: Logger,
	chainModule: Chain,
	bftModule: BFT,
	processorModule: Processor,
): Promise<void> => {
	// Get all blocks and find lowest height (next one to be applied), as it should return in height desc
	const tempBlocks = await chainModule.dataAccess.getTempBlocks();
	const blockLowestHeight = tempBlocks[tempBlocks.length - 1];
	const blockHighestHeight = tempBlocks[0];

	const forkStatus = bftModule.forkChoice(blockHighestHeight.header, chainModule.lastBlock.header);
	const blockHasPriority =
		forkStatus === ForkStatus.DIFFERENT_CHAIN || forkStatus === ForkStatus.VALID_BLOCK;

	// Block in the temp table has preference over current tip of the chain
	if (blockHasPriority) {
		logger.info('Restoring blocks from temporary table');
		await deleteBlocksAfterHeight(
			processorModule,
			chainModule,
			logger,
			blockLowestHeight.header.height - 1,
			false,
		);
		// In case fork status is DIFFERENT_CHAIN - try to apply blocks from temp_blocks table
		await restoreBlocks(chainModule, processorModule);
		logger.info('Chain successfully restored');
	} else {
		// Not different chain - Delete remaining blocks from temp_blocks table
		await clearBlocksTempTable(chainModule);
	}
};

export const computeBlockHeightsList = (
	finalizedHeight: number,
	activeDelegates: number,
	listSizeLimit: number,
	currentRound: number,
): number[] => {
	const startingHeight = Math.max((currentRound - 1) * activeDelegates, 0);
	const heightList = new Array(listSizeLimit)
		.fill(0)
		.map((_, i) => startingHeight - i * activeDelegates)
		.filter(height => height >= 0);
	const heightListAfterFinalized = heightList.filter(height => height > finalizedHeight);
	return heightList.length !== heightListAfterFinalized.length
		? [...heightListAfterFinalized, finalizedHeight]
		: heightListAfterFinalized;
};

export const computeLargestSubsetMaxBy = <T>(
	arrayOfObjects: T[],
	propertySelectorFunc: (param: T) => number,
): T[] => {
	const comparableValues = arrayOfObjects.map(propertySelectorFunc);
	const absoluteMax = Math.max(...comparableValues);
	const largestSubset = [];
	// eslint-disable-next-line no-restricted-syntax
	for (const item of arrayOfObjects) {
		if (propertySelectorFunc(item) === absoluteMax) {
			largestSubset.push(item);
		}
	}
	return largestSubset;
};
