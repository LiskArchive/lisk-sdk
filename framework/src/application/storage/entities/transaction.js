/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const _ = require('lodash');
const path = require('path');
const {
	entities: { Transaction: TransactionEntity },
} = require('../../../components/storage');

const sqlFiles = {
	create: 'transactions/create.sql',
};

const trsCreateFields = [
	'id',
	'blockId',
	'type',
	'timestamp',
	'senderPublicKey',
	'senderId',
	'recipientId',
	'amount',
	'fee',
	'signature',
	'signSignature',
	'signatures',
	'asset',
	'transferData',
];

class ChainTransaction extends TransactionEntity {
	constructor(adapter, defaultFilters = {}) {
		super(adapter, defaultFilters);

		this.sqlDirectory = path.join(path.dirname(__filename), '../sql');

		this.SQLs = this.loadSQLFiles('transaction', sqlFiles, this.sqlDirectory);
	}

	create(data, _options, tx) {
		const transactions = ChainTransaction._sanitizeCreateData(data);

		const createSet = this.getValuesSet(transactions, trsCreateFields);

		return this.adapter.executeFile(
			this.SQLs.create,
			{ values: createSet, attributes: trsCreateFields },
			{ expectedResultCount: 0 },
			tx,
		);
	}

	static _sanitizeCreateData(data) {
		const transactions = Array.isArray(data)
			? _.cloneDeep(data)
			: [_.cloneDeep(data)];

		const recipientTransactionTypes = [0, 3, 8];

		transactions.forEach(transaction => {
			transaction.signatures = transaction.signatures
				? transaction.signatures.join()
				: null;

			if (recipientTransactionTypes.includes(transaction.type)) {
				transaction.amount = transaction.asset.amount.toString();
				transaction.recipientId = transaction.asset.recipientId;
			} else {
				transaction.recipientId = null;
				transaction.amount = 0;
			}

			transaction.fee = transaction.fee.toString();
			transaction.transferData = null;

			// Transfer data is bytea and can not be included as json when null byte is present
			const dataTransactionType = [0, 8];
			if (
				dataTransactionType.includes(transaction.type) &&
				transaction.asset &&
				transaction.asset.data
			) {
				transaction.transferData = Buffer.from(transaction.asset.data, 'utf8');
				delete transaction.asset;
			}

			// stringify should be done after converting asset.data into transferData
			transaction.asset = transaction.asset
				? JSON.stringify(transaction.asset)
				: null;
		});

		return transactions;
	}
}

module.exports = ChainTransaction;
