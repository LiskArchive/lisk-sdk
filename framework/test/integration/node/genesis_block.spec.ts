/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { KVStore } from '@liskhq/lisk-db';
import { Account } from '@liskhq/lisk-chain';
import { validator } from '@liskhq/lisk-validator';
import { nodeUtils } from '../../utils';
import { createDB, removeDB } from '../../utils/kv_store';
import { Node } from '../../../src/node';
import { genesis, genesisBlock as getGenesisBlock, DefaultAccountProps } from '../../fixtures';
import { createTransferTransaction } from '../../utils/node/transaction';

describe('genesis block', () => {
	const dbName = 'genesis_block';
	const genesisBlock = getGenesisBlock();
	let node: Node;
	let blockchainDB: KVStore;
	let forgerDB: KVStore;

	beforeAll(async () => {
		({ blockchainDB, forgerDB } = createDB(dbName));
		node = await nodeUtils.createAndLoadNode(blockchainDB, forgerDB);
		// Since node start the forging so we have to stop the job
		// Our test make use of manual forging of blocks
		node['_forgingJob'].stop();
	});

	afterAll(async () => {
		await node.cleanup();
		await blockchainDB.close();
		await forgerDB.close();
		removeDB(dbName);
	});

	describe('given the application has not been initialized', () => {
		describe('when chain module is bootstrapped', () => {
			it('should save genesis block to the database', async () => {
				const block = await node['_chain'].dataAccess.getBlockByID(genesisBlock.header.id);

				expect(block.header.version).toEqual(0);
				expect(block.header).toEqual(genesisBlock.header);
			});

			it('should save accounts from genesis block assets', async () => {
				// Get genesis accounts
				const genesisAccounts = genesisBlock.header.asset.accounts;

				// Get delegate accounts in genesis block from the database
				const accountsFromDb = await Promise.all(
					genesisAccounts.map(async account =>
						node['_chain'].dataAccess.getAccountByAddress(account.address),
					),
				);

				expect(genesisAccounts).toEqual(accountsFromDb);
			});

			it('should have correct delegate list', async () => {
				const delegateListFromChain = await nodeUtils.getDelegateList(node);
				expect(delegateListFromChain).toMatchSnapshot();
			});
		});
	});

	describe('given the application was initialized earlier', () => {
		const account = (genesisBlock.header.asset.accounts[
			genesisBlock.header.asset.accounts.length - 1
		] as unknown) as Account<DefaultAccountProps>;
		let newBalance: bigint;
		let oldBalance: bigint;

		beforeEach(async () => {
			// FIXME: Remove with #5572
			validator.removeSchema('/block/header');
			const genesisAccount = await node['_chain'].dataAccess.getAccountByAddress<
				DefaultAccountProps
			>(genesis.address);
			const recipient = await node['_chain'].dataAccess.getAccountByAddress<DefaultAccountProps>(
				account.address,
			);
			oldBalance = account.token.balance;
			newBalance = oldBalance + BigInt('100000000000');

			const transaction = createTransferTransaction({
				amount: BigInt('100000000000'),
				recipientAddress: recipient.address,
				networkIdentifier: node['_networkIdentifier'],
				nonce: genesisAccount.sequence.nonce,
				passphrase: genesis.passphrase,
			});
			const newBlock = await nodeUtils.createBlock(node, [transaction]);
			await node['_processor'].process(newBlock);
		});

		describe('when chain module is bootstrapped', () => {
			it('should not apply the genesis block again', async () => {
				// Act
				// Re-initialize the node
				node = await nodeUtils.createAndLoadNode(blockchainDB, forgerDB);

				// Arrange & Assert
				const recipient = await node['_chain'].dataAccess.getAccountByAddress<DefaultAccountProps>(
					account.address,
				);
				expect(recipient.token.balance).toEqual(newBalance);
			});
		});
	});
});
