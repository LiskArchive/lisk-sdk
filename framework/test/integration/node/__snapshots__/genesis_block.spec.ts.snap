// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genesis block given the application has not been initialized when chain module is bootstrapped should have correct delegate list 1`] = `
Array [
  "8f6d2a3adc1d5250108f3207a25519955d90991f",
  "8f8fe4d97de5e36fd626f68b5d5b11da1f867635",
  "8ffd6d8167307044edaa5d61e5d89c5abdba41e8",
  "92662e3f51c1c2616681602f8f2efe6540d08789",
  "97008cacc691168b13c2e38eb645360cffc9c03b",
  "983528b238517c93307546b7282ffd574bb111de",
  "9b82f132f9a2ff243c4f5e2a363577158fb40f09",
  "9d9d11db10cba9d06ec1ca40b8722b5b71d442c1",
  "9e4da082aa7ded71eb20a31a6f7de40ed9dfdae3",
  "a1e0b9297715e31097d64c2cf931ae0ac1ac9a9b",
  "a35704117c4bc7d14282bd9893c79add67fe973b",
  "a47fce35acb7a9301772ab043fd5b6d56bd4b5d0",
  "aa12e60addf6601b6b4928e9abe39b8ed9f3115f",
  "aa29d53de5361e4f226d1c377e96fbfae4fe9163",
  "ae7905696771a06123da09e7a782beb35b2dfd7b",
  "af711c5fecdc10bd1546b74860518586bef42fc3",
  "b511353e8d4df37ad688ca0bb43965298cc37d69",
  "b858435ecf44fd9c5f007be507da8faf777cb2bc",
  "bae232c8f2689aea9a12577786fa3141f561171c",
  "c391dde78f07171e191f8d6ba17d053873154998",
  "c3ad0b22afcb9e8e948624283452bd181dcd04c3",
  "c721ffa7c7a7cf11e13096fddc6af24974a4ee8e",
  "cb4f44724bab013876473dfd65b87448bc4cd381",
  "cd2764890d47e7aa0cdf15501b11c028a6e3aca0",
  "ce07270e060615b80795adcf3a6526bfef3f5ceb",
  "d0f26e13c229e5d7c745ef0545bb4fb0f759e90e",
  "d4bfee674d97f5b8724bfa1e1ddf8a55814cd341",
  "db07dec164e52d22ddd0018fa17dde7a1c36732b",
  "dd8386f4813122fc836f269de6d0a07c15f49583",
  "ddbc46e77f8f183bf99fe3206c570c54da9e398a",
  "dec7cd87252110c02fc681c81fad29cfa9f9231e",
  "e4037d63d71c28bb144ff39c9de68a3060d5e631",
  "e5bc74611fca712101f6a0d193725b336f550566",
  "e74aa95a04a429d9d417584d90b325f3993068e9",
  "e97bcae362ebba20268a264a59a96ed4c79ba551",
  "ea5b7a4aa1b8bea278cfad3f5304606344dba4e5",
  "ed0fc1cb6dc8e27ecbedb0b6faeede345db47837",
  "ee531c16f419b1fce5a86150ef5d755d90a062bb",
  "f0ad62b4f92e55db0943f4bb10645f213e0eab41",
  "f42fe50d8008a4b30d4d2ac510e083c6cf1d6615",
  "f598441241d79b620ccf61f0907291a08d460b66",
  "f8a7fc8d5fdbf0dfd921202ed7e4b35350754b26",
  "fccf1a0f5532234118250e894b299eacc0ed8c24",
  "ff52bc34f842b9102cb17f2d3e755b2414b3f70d",
  "02793f74342b4d58174b4387f54252c80a605e66",
  "04b004136b2839851eebe8cfba3c8539eeb74452",
  "0d4252621f47895c6f4a43e105e68b344b7d5d34",
  "0d4f07890f8b89bd9cd852fd92faef8130a82d3e",
  "0f8e8b9df5245cccbe3404a3be0baae4876e333c",
  "11e8aebf3f0da4b49dd42392648dd06d46e4e459",
  "1348bdced23cbdfb92cf3c74742d8f3d96f436de",
  "14785e33717634c91d63349ae1c0366c072be25c",
  "14a0751c71e2871b5d88ed060c93a72a3834ba85",
  "150fbd209c81c95de1e0eb3399baa0cf9d2853ed",
  "173129339f6b56e72345fd6529a552db6ea87f15",
  "18d0b82999dfab59e7a7d6592e9f2973798799cb",
  "1aa54d505856beb417f706889d16cd36f677a4c8",
  "1bea5df44a2bfa8b864cddfb6138c71a883895ed",
  "1cc0e48484122141e38098e5083961cc8e430dca",
  "1d7be48ed2d30cc901b1a1ee636f57b0c560878c",
  "1dc798f399d525f9e8cc1ff3d566681c4f62b867",
  "1e0652d45c9a2d37a53c40d0e060417c3c15d126",
  "1e964184beb1d146f5a1e8de2e50341933bcbc5a",
  "229ffdf682c09ffeb3bcf2d94e13467bd9a50861",
  "244ce8c5bd9744aeff8a552bb150e56af78ac6b8",
  "24645f4c3b1f536bb8c018528218d87120505edb",
  "29a90d7c09c8eb26e505a684b5681785e76794ad",
  "2e63c4d179887f3612c878928f4327a5d2649246",
  "2f7f16feef2241085b6bca3ba877969a90e3f725",
  "329a31a05227a3c07bac32b1b73b89d2432dd934",
  "329bc1cbfdd588b92f66834d8bf3f02c8da4ccf1",
  "3376eddd20564b9b45d630e7ce261311f0e230d9",
  "3801ed0f31990a97b2c0a70da1ad2bb4aa1ea93d",
  "39e43a8de34509a45c39e5ff338601567077b3c4",
  "3c4742625f4ff24a4dcd5a608ec43e23e8a799ca",
  "44111c7045776d79e25df5d2cc3e15313516ffb6",
  "4eb5eca5c4b0301804ac2c6c9879902f9584082c",
  "54f82c814b4524f1d3937cc4105b02b288e950bc",
  "5770ebefba0befa6e2eecad5a6d4f12b86cccfe7",
  "589ca3fda0680e234473e4c44a84341f5eb1e752",
  "58a01618f3a986c75aec5266b9c52722a1eec0d3",
  "5dafaad782ef1e95c95d1a05ff7b3ac8b7a9d7a6",
  "62bf33222be25d000bda5cc4e671f5d7db68eddf",
  "668769cc41b7d8bbcc82166d479bc27b3bf88677",
  "66b49e0dd08e8152ed2d1ea30ebaae3bb0284340",
  "66e0235892d17ddc6cf39862aa3ee64ec27f9bd1",
  "66e06bdd6bcfac4eea7bdb79414fdec70efa1904",
  "6bbfefd0c6ef251f63e42c430e1b98255f682c4b",
  "6c955cf30dd41f1a6d8590b223043d91b41f890b",
  "6d7be6b821ff082a5de312c49717262a8859464b",
  "7b30ef13a16a61e9b560b1f9fadbc5de1ba16360",
  "7e193c5e39092eb1b1f81ccded6f700652856d10",
  "8415c1ac3b53f34039b1497589d7dcfaa11d41a9",
  "84a7d87a34bac987b0c0d708e782ec27e1aba7f2",
  "85f0e7a39531876e89c4d5eeab9d98adbbbaaf88",
  "87a60bb1d1551b5ba17eabc220a8c6ad09dc1e87",
  "8a00e7838582b4e432881d98952494837d903671",
  "8aefb8f0a34bf10fdc1720c283fb73ea1ba117c6",
  "8c66f23d15a9bf14c948ec6c060a0afbdec4246f",
  "8d5cb3bd72d97bba55c980750fdfe0399c96dfc1",
  "8e561cdaab90d883994b5f1582105894453f5cc8",
]
`;
